engine/jobmaster/dm/checkpoint/agent.go

```diff
@@ -156,6 +168,99 @@ func (c *AgentImpl) UpgradeFuncs() []bootstrap.UpgradeFunc {
 	return nil
 }
 
+// FetchAllDoTables returns all need to do tables after filtered and routed (fetches from upstream MySQL).
+func (c *AgentImpl) FetchAllDoTables(ctx context.Context, cfg *config.JobCfg) (map[metadata.TargetTable][]metadata.SourceTable, error) {
+	c.logger.Info("fetch all do tables from upstream")
+
+	var (
+		mu      sync.Mutex
+		result  = make(map[metadata.TargetTable][]metadata.SourceTable, len(cfg.Upstreams))
+		g, gCtx = errgroup.WithContext(ctx)
+	)
+
+	for _, upstream := range cfg.Upstreams {
+		up := upstream
+		// filter/route in errgroup will cause data race.
+		baList, err := filter.New(up.CaseSensitive, cfg.BAList[up.BAListName])
+		if err != nil {
+			return result, err
+		}
+		routeRules := make([]*router.TableRule, 0, len(up.RouteRules))
+		for _, ruleName := range up.RouteRules {
+			routeRules = append(routeRules, cfg.Routes[ruleName])
+		}
+		router, err := regexprrouter.NewRegExprRouter(up.CaseSensitive, routeRules)
+		if err != nil {
+			return result, err
+		}
+
+		g.Go(func() error {
+			db, err := conn.DefaultDBProvider.Apply(conn.UpstreamDBConfig(up.DBCfg))
+			if err != nil {
+				return err
+			}
+			defer db.Close()
+
+			// fetch all do tables
+			sourceTables, err := conn.FetchAllDoTables(gCtx, db, baList)
+			if err != nil {
+				return err
+			}
+
+			for schema, tables := range sourceTables {
+				for _, table := range tables {
+					targetSchema, targetTable, err := router.Route(schema, table)
+					if err != nil {
+						return err
+					}
+
+					target := metadata.TargetTable{Schema: targetSchema, Table: targetTable}
+					mu.Lock()
+					result[target] = append(result[target], metadata.SourceTable{Source: up.SourceID, Schema: schema, Table: table})
+					mu.Unlock()
+				}
+			}
+			return nil
+		})
+
+	}
+	err := g.Wait()
+	return result, err
+}
+
+// FetchTableStmt fetch create table statement from checkpoint.
```

Comment:
the comments can be moved or repeated to interface, so mouse hovering on invocations will see it.